### PR TITLE
Cipher.v: state/key/add_round_key

### PR DESCRIPTION
**Pull request by smore-lore**
This uses Vector.t to implement state and key, and it adds a
definition of add_round_key and proves it is its own inverse.

- suggestions for using VectorUtils.v appreciated (export vnil?)
- Adding nc/nb to context instead of making them dependently typed was causing some issues in my proofs where
  it couldn't match 'nc' and 'nc0' with each other.

### CODE REVIEW

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (satnam6502) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:21]: This is a vector of bool so you could write:
```coq
Vector.t bool nb
```
Even better would be to import the `Bvector` package and write:
```coq
From Coq Require Import Bool.Bvector.

Definition key (nb : nat) := Bvector nb.
```
Or just make a type synonym for `key` where you need it e.g. using `Notation`.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (satnam6502) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:23]: I suspect this is in one of the standard libraries that has a `Bvector` instance.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (satnam6502) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:27]: I would make `nb` an implicit argument.
```coq
Definition add_round_key nc {nb : nat} (input : state nc nb) (k : key nb) :
```

**Comment by satnam6502** [comment]
Thanks! That's a great start! I'll defer to the others to complete this review.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v]: Thanks, done, nicer.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v]: Done

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v]: Nice suggestion, thanks & done.

**Comment by satnam6502** [comment]
Thanks for putting several of us down as reviewers. I propose that since @jadephilipoom  is leading the space work as soon as she gives GitHib Approval it is fine to merge if that's fine with @cipher1024 and @blaxill 
I don't think we have a policy on what to do with PRs with multiple reviewers?

**Review by satnam6502 — APPROVED** [review]

**Comment by smore-lore** [comment]
> I don't think we have a policy on what to do with PRs with multiple reviewers?

In general - I like multiple reviewers or being included on changes even if I'm not necessarily the one to approve, since reading a lot of code gives you context for what is going on. I expected anyone with access and who was comfortable with my change to approve and merge without necessarily everyone reviewing.

**Comment by blaxill** [comment]
> I don't think we have a policy on what to do with PRs with multiple reviewers?

The SWE team has a policy that one review is required before merging, but you can request as many as you like to get feedback. I think we should go with that.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (blaxill) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:38]: (Your approach is fine) Another approach that can sometimes work in the face of `destruct` failing due to dependent types is to refine with a match on the variable:

```Coq
    Goal forall b,
     BVxor 0 (BVxor 0 [] b) b = [].
    Proof.
      intros.
            intros.
      refine match b with
             | nil _ => _
             end; auto.
    Qed.
```

**Review by blaxill — APPROVED** [review]

**Comment by cipher1024** [comment]
Because `BVxor` is defined in terms of `map`, I was suggesting to @smore-lore to use the functorial laws to reduce the proof 
to the property that `xor` on single bits is idempotent. What is your opinion on this approach @jadephilipoom, @blaxill and @satnam6502?

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (blaxill) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:40]: (Again totally optional, might be useful for future proofs) Some alternative ways to write this :

```Coq
        rewrite !map2_cons. (* or *)
        rewrite ?map2_cons. (* or *)
        do 2 rewrite map2_cons. (* or *)
        repeat rewrite map2_cons. 
```

**Comment by satnam6502** [comment]
> Because `BVxor` is defined in terms of `map`, I was suggesting to @smore-lore to use the functorial laws to reduce the proof
> to the property that `xor` on single bits is idempotent. What is your opinion on this approach @jadephilipoom, @blaxill and @satnam6502?

Sure, you could `unfold BVxor` and see if you can apply the `map2_*` lemmas?

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (jadephilipoom) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:21]: Thanks for taking the suggestion to move this to its own file. Just to respond to your last comment on the topic [(discussion from the earlier post of this change)](https://github.com/smore-lore/oak-hardware/commit/09f910d7c16f9a90f3e38b87c8bc73b29e330e88#r43689206):

> This makes sense - I'm curious how exactly we keep all the types aligned then? If one file declares a key which is incompatible with another file, we'd only find out in a 3rd file which tries to use the functionality together?
> 
> Should there be some basic type outline we check all these files against?

Yes, there will be files that thread these pieces together and will fail if they are incompatible -- the places where we prove that some implementation is equivalent to the specification will plug type arguments into the specification, and rely on proofs that plug type arguments into any subroutine specifications.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (jadephilipoom) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:20]: `nc` and `nb` could probably be declared with `Context` here, instead of being passed individually to each definition. `Context` will insert the variable as an argument only for definitions within the section that _use_ the variable, so the type signatures here would not be changed.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (jadephilipoom) [inline silveroak-opentitan/aes/Spec/AddRoundKey.v:54]: It's completely fine as-is, but when I have a very easily-solved induction base case I often save myself curly braces this way:
```suggestion
      induction x; [ reflexivity | ].
```

The `[ reflexivity | ]` syntax means "expect two goals, and run `reflexivity` on the first one".

**Review by jadephilipoom — APPROVED** [review]
Great work! Thanks for putting this together.

**Comment by jadephilipoom** [comment]
> Adding nc/nb to context instead of making them dependently typed was causing some issues in my proofs where
it couldn't match 'nc' and 'nc0' with each other.

I recommended this change in my review, and am curious where issues popped up. Is it possible that the proof introduced another `nc`? Proofs that use context variables will end up forall-quantified over those variables, so writing a proof that says `forall nc` within a section that already declared `nc` will insert two competing versions of `nc` and possibly produce the issue you described.

I see that in `inverse_add_round_key` you do induction on a `state`, which could cause issues with the `nc` variable not matching up as well. If that's where you're running into issues, feel free to ignore the suggestion and leave as-is!

**Comment by jadephilipoom** [comment]
> Because `BVxor` is defined in terms of `map`, I was suggesting to @smore-lore to use the functorial laws to reduce the proof
> to the property that `xor` on single bits is idempotent. What is your opinion on this approach @jadephilipoom, @blaxill and @satnam6502?

I think this is a great approach! And actually, I believe you could now move the proofs that are just about bitvectors to `cava/Cava/BitArithmetic.v`, the file for generic proofs about bit vectors, so as to minimize complexity in `Spec`.

**Comment by smore-lore** [comment]
re jadephilipoom - this is the error I get when I move nc/nb into a context in inverse_add_round_key when I try to do induction on x. Couldn't figure out why I was introducing another nc0, so this is why I made it dependently typed. It's probably a limitation of my coq knowledge. Perhaps we could change it when/if it becomes more important, unless you want to help out!

```
Abstracting over the terms "nc" and "x" leads to a term
fun (nc0 : nat) (x0 : t (Bvector nb) nc0) =>
map (fun col : Vector.t bool nb => BVxor nb col k)
  (map (fun col : Vector.t bool nb => BVxor nb col k) x0) = x0
which is ill-typed.
Reason is: Illegal application: 
The term "@eq" of type "forall A : Type, A -> A -> Prop"
cannot be applied to the terms
 "state" : "Set"
 "map (fun col : Vector.t bool nb => BVxor nb col k)
    (map (fun col : Vector.t bool nb => BVxor nb col k) x0)" : "t (Vector.t bool nb) nc0"
 "x0" : "t (Bvector nb) nc0"
The 2nd term has type "t (Vector.t bool nb) nc0" which should be coercible to 
"state".
```

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v:38]: Thanks for the tip!

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v]: Done.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v:21]: Perfect, that makes sense.

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v]: Done, thanks for the tip.

**Comment by smore-lore** [comment]
> I think this is a great approach! And actually, I believe you could now move the proofs that are just about bitvectors to `cava/Cava/BitArithmetic.v`, the file for generic proofs about bit vectors, so as to mi

These proofs are pretty specific to this operation, and I didn't want to carve out a whole new section there for inverse_xorb. Let me keep them here for now, and if they are needed by something else, we can move them. I'm also worried I may want to change the key types used here after we start looking at the other operations.

**Comment by jadephilipoom** [comment]
> re jadephilipoom - this is the error I get when I move nc/nb into a context in inverse_add_round_key when I try to do induction on x. Couldn't figure out why I was introducing another nc0, so this is why I made it dependently typed.

The issue is that the context variable "fixes" the size of the vector, so doing induction doesn't work because the proof doesn't have a notion of a `state` with any other length. Try putting in `cbv [state] in *` just before doing induction, so `x` is seen as `Vector.t` instead of `state`?

File: silveroak-opentitan/aes/Spec/AddRoundKey.v
Comment (smore-lore) [reply silveroak-opentitan/aes/Spec/AddRoundKey.v]: done, w/ your help

**Comment by smore-lore** [comment]
I think this is ready to merge as discussed (if someone can - I don't have perms). Will talk to Simon about cleaning up these proofs in a separate PR.

**Pull request merged**
